packages/test/src/integration-tests/workflows.ts

```diff
@@ -0,0 +1,74 @@
+import * as wf from '@temporalio/workflow';
+
+export const workflowsPath = __filename;
+
+// An update with arguments and return value, with which we associate an async
+// handler function and a validator.
+export const update = wf.defineUpdate<string[], [string]>('update');
+// A update that accepts no arguments and returns nothing, with which we
+// associate a sync handler function, but no validator.
+export const doneUpdate = wf.defineUpdate<void, []>('done-update');
+
+export async function workflowWithUpdates() {
+  let state: string[] = [];
+  const updateHandler = async (arg: string): Promise<string[]> => {
+    state.push(arg);
+    if (arg == 'fail-update') {
+      throw new wf.ApplicationFailure(`Deliberate ApplicationFailure in handler`);
+    }
+    return state;
+  };
+  const doneUpdateHandler = (): void => {
+    state.push('done');
+  };
+  const validator = (arg: string): void => {
+    if (arg == 'bad-arg') {
+      throw new Error('Validation failed');
+    }
+  };
+  wf.setHandler(update, updateHandler, { validator });
+  wf.setHandler(doneUpdate, doneUpdateHandler);
+  await wf.condition(() => state.includes('done'));
+  state.push('$');
+  return state;
+}
+
+export const updateWithMutableArg = wf.defineUpdate<string[], [[string]]>('updateWithMutableArg');
+
+export async function workflowWithMutatingValidator() {
+  let state: string[] = [];
+  const updateHandler = async (arg: [string]): Promise<string[]> => {
+    state.push(arg[0]);
+    return state;
+  };
+  const doneUpdateHandler = (): void => {
+    state.push('done');
+  };
+  const validator = (arg: [string]): void => {
+    arg[0] = 'mutated!';
+  };
+  wf.setHandler(updateWithMutableArg, updateHandler, { validator });
+  wf.setHandler(doneUpdate, doneUpdateHandler);
+  await wf.condition(() => state.includes('done'));
+  state.push('$');
+  return state;
+}
+
+/* BEGIN: Test example from WorkflowHandle docstring */
+export const incrementSignal = wf.defineSignal<[number]>('increment');
+export const getValueQuery = wf.defineQuery<number>('getValue');
+export const incrementAndGetValueUpdate = wf.defineUpdate<number, [number]>('incrementAndGetValue');
+
+export async function counterWorkflow(initialValue: number): Promise<void> {
+  let count = initialValue;
+  wf.setHandler(incrementSignal, (arg: number) => {
+    count += arg;
+  });
+  wf.setHandler(getValueQuery, () => count);
+  wf.setHandler(incrementAndGetValueUpdate, (arg: number): number => {
```

Comment:
Updates should always be async functions IMHO, it simplifies the signatures for handles.

Reply:
Hrmm, this would simplify the types if it _always_ had to return `Promise<T>` instead of `T | Promise<T>`

Reply:
As mentioned elsewhere my current inclination is to support normal (non-async) functions as Update handlers, unless it causes severe technical problems or un-ergonomic public interfaces. AFAIK it's a perfectly valid usage of Update to synchronously mutate and return internal workflow state without doing anything that needs to be async. So it seems intellectually confusing to me to demand that they are async. However, please let me know if that design position is questionable; I wasn't around when Update was conceived of.